File: 7_colas_de_prioridad/k_closest_points_to_origin.py
# Programando K Closest Points to Origin con Python
# no hay tipo nativo para colas de prioridad pero usamos una libreria heappq:
import heapq
from typing import List


def kClosest(puntos: List[List[int]], K: int) -> List[List[int]]:
    if K == len(puntos):
        return puntos
    distancias = []
    for x, y in puntos:
        distancia = (x*x + y*y)**(1/2)
        distancias.append((distancia, [x, y]))
    heapq.heapify(distancias)
    puntosMasCercanos = []
    for _ in range(K):
        distanciaActual, coordenadas = heapq.heappop(distancias)  # log
        puntosMasCercanos.append(coordenadas)
    return puntosMasCercanos


# Example usage
result = kClosest([[1, 3], [3, 4], [5, 6]], 2)
print(result)  # [[1,3]]

""" The code above does the following, explained in English:
1. If the number of points in the list is equal to K, we just return the entire list.
2. We create an empty list called distances.
3. We loop through all the points and calculate the distance of each point from the origin and store the distance and the coordinates of the point in the list distances.
4. We then convert the list distances into a min heap.
5. We then pop the first K entries from the heap and return the list of the coordinates of the points. """

File: 7_colas_de_prioridad/k_closest_points_to_origin.ts
// # Programando K Closest Points to Origin con Python

function kClosest(puntos, K) {
  if (K === puntos.length) {
    return puntos;
  }
  let distancias = [];
  for (let [x, y] of puntos) {
    let distancia = Math.sqrt(x * x + y * y);
    distancias.push([distancia, [x, y]]);
  }
  distancias.sort((a, b) => a[0] - b[0]);
  let puntosMasCercanos = [];
  for (let i = 0; i < K; i++) {
    let [distanciaActual, coordenadas] = distancias[i];
    puntosMasCercanos.push(coordenadas);
  }
  return puntosMasCercanos;
}

console.log(
  kClosest(
    [
      [1, 3],
      [3, 4],
      [5, 6],
    ],
    1,
  ),
);
console.log(
  kClosest(
    [
      [13, -8],
      [1, 20],
      [-5, 5],
    ],
    2,
  ),
);
